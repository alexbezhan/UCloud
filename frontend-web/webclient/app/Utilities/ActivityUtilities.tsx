import { buildQueryString } from "./URIUtilities";
import { ScrollRequest } from "Scroll";
import { ActivityFilter } from "Activity";

export const activityQuery = (scroll: ScrollRequest<number>, filter?: ActivityFilter) => {
    let params: any = {};
    if (!!scroll.offset) params.offset = scroll.offset;
    params.scrollSize = scroll.scrollSize;

    if (filter !== undefined) {
        if (filter.type !== undefined) params.type = filter.type;
        if (filter.collapseAt !== undefined) params.collapseAt = filter.collapseAt;
        if (filter.maxTimestamp !== undefined) params.maxTimestamp = filter.maxTimestamp.getTime();
        if (filter.minTimestamp !== undefined) params.minTimestamp = filter.minTimestamp.getTime();
    }

    return buildQueryString("/activity/browse/user", params);
};

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}`;

export const getLatestActivity = () => undefined;