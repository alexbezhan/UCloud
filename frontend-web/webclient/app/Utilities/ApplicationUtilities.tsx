import {
    ApplicationInvocationDescription,
    ApplicationMetadata,
    ApplicationParameter,
    JobState,
    ParameterTypes,
    RunsSortBy
} from "Applications";
import {RangeRef} from "Applications/Widgets/RangeParameters";
import Cloud from "Authentication/lib";
import {SortOrder} from "Files";
import * as React from "react";
import {snackbarStore} from "Snackbar/SnackbarStore";
import {Page} from "Types";
import {addStandardDialog} from "UtilityComponents";
import {errorMessageOrDefault, removeTrailingSlash} from "UtilityFunctions";
import {expandHomeFolder} from "./FileUtilities";

export const hpcJobQueryPost = "/hpc/jobs";

export const hpcJobQuery = (id: string) => `/hpc/jobs/${encodeURIComponent(id)}`;

export function hpcJobsQuery(
    itemsPerPage: number,
    page: number,
    sortOrder?: SortOrder,
    sortBy?: RunsSortBy,
    minTimestamp?: number,
    maxTimestamp?: number,
    filter?: JobState
): string {
    let query = `/hpc/jobs/?itemsPerPage=${itemsPerPage}&page=${page}`;
    if (sortOrder) query = query.concat(`&order=${sortOrder}`);
    if (sortBy) query = query.concat(`&sortBy=${sortBy}`);
    if (minTimestamp != null) query = query.concat(`&minTimestamp=${minTimestamp}`);
    if (maxTimestamp != null) query = query.concat(`&maxTimestamp=${maxTimestamp}`);
    if (filter != null) query = query.concat(`&filter=${filter}`);
    return query;
}

export function advancedSearchQuery(): string {
    return "/hpc/apps/advancedSearch";
}

export const hpcFavoriteApp = (name: string, version: string) =>
    `/hpc/apps/favorites/${encodeURIComponent(name)}/${encodeURIComponent(version)}`;

export const hpcFavorites = (itemsPerPage: number, pageNumber: number) =>
    `/hpc/apps/favorites?itemsPerPage=${itemsPerPage}&page=${pageNumber}`;

export const hpcApplicationsQuery = (page: number, itemsPerPage: number) =>
    `/hpc/apps?page=${page}&itemsPerPage=${itemsPerPage}`;

interface HPCApplicationsSearchQuery {
    query: string;
    page: number;
    itemsPerPage: number;
}

export const hpcApplicationsTagSearchQuery = ({query, page, itemsPerPage}: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/searchTags?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const cancelJobQuery = `hpc/jobs`;

export const cancelJobDialog = (
    {jobId, onConfirm, jobCount = 1}: {
        jobCount?: number,
        jobId: string,
        onConfirm: () => void
    }
): void =>
    addStandardDialog({
        title: `Cancel job${jobCount > 1 ? "s" : ""}?`,
        message: jobCount === 1 ? `Cancel job: ${jobId}?` : "Cancel jobs?",
        cancelText: "No",
        confirmText: `Cancel job${jobCount > 1 ? "s" : ""}`,
        onConfirm
    });

export const cancelJob = (cloud: Cloud, jobId: string): Promise<{request: XMLHttpRequest, response: void}> =>
    cloud.delete(cancelJobQuery, {jobId});

interface FavoriteApplicationFromPage<T> {
    name: string;
    version: string;
    page: Page<{metadata: ApplicationMetadata, favorite: boolean} & T>;
    cloud: Cloud;
}

/**
 * Favorites an application.
 * @param {Application} Application the application to be favorited
 * @param {Cloud} cloud The cloud instance for requests
 */
export async function favoriteApplicationFromPage<T>(
    {
        name,
        version,
        page,
        cloud
    }: FavoriteApplicationFromPage<T>
): Promise<Page<T>> {
    const a = page.items.find(it => it.metadata.name === name && it.metadata.version === version)!;
    try {
        await cloud.post(hpcFavoriteApp(name, version));
        a.favorite = !a.favorite;
    } catch (e) {
        snackbarStore.addFailure(errorMessageOrDefault(e, `An error ocurred favoriting ${name}`));
    }
    return page;
}


interface StringMap {
    [k: string]: string;
}

interface AllowedParameterKey {
    name: string;
    type: ParameterTypes;
}

interface ExtractParameters {
    nameToValue: StringMap;
    allowedParameterKeys: AllowedParameterKey[];
    siteVersion: number;
}

export const findKnownParameterValues = (
    {
        nameToValue,
        allowedParameterKeys,
        siteVersion
    }: ExtractParameters
): StringMap => {
    const extractedParameters = {};
    if (siteVersion === 1) {
        allowedParameterKeys.forEach(({name, type}) => {
            if (nameToValue[name] !== undefined) {
                if (typeMatchesValue(type, nameToValue[name])) {
                    extractedParameters[name] = nameToValue[name];
                }
            }
        });
    }
    return extractedParameters;
};

export const isFileOrDirectoryParam = ({type}: {type: string}) => type === "input_file" || type === "input_directory";

const typeMatchesValue = (type: ParameterTypes, parameter: string | [number, number]): boolean => {
    switch (type) {
        case ParameterTypes.Boolean:
            return parameter === "Yes" || parameter === "No" || parameter === "";
        case ParameterTypes.Integer:
            return parseInt(parameter as string, 10) % 1 === 0;
        case ParameterTypes.FloatingPoint:
            return typeof parseFloat(parameter as string) === "number";
        case ParameterTypes.Range:
            return typeof parameter === "object" && "size" in parameter;
        case ParameterTypes.Text:
        case ParameterTypes.InputDirectory:
        case ParameterTypes.InputFile:
        case ParameterTypes.SharedFileSystem:
        case ParameterTypes.Peer:
            return typeof parameter === "string";

    }
};

interface ExtractedParameters {
    [key: string]: string | number | boolean |
    {source: string, destination: string;} |
    {min: number, max: number} |
    {fileSystemId: string} |
    {jobId: string};
}

export type ParameterValues = Map<string, React.RefObject<HTMLInputElement | HTMLSelectElement | RangeRef>>;

interface ExtractParametersFromMap {
    map: ParameterValues;
    appParameters: ApplicationParameter[];
    cloud: Cloud;
}

export function extractValuesFromWidgets({map, appParameters, cloud}: ExtractParametersFromMap): ExtractedParameters {
    const extracted: ExtractedParameters = {};
    map.forEach((r, key) => {
        const parameter = appParameters.find(it => it.name === key);
        if (!r.current) return;
        if (("value" in r.current && !r.current.value) || ("checkValidity" in r.current && !r.current.checkValidity())) return;
        if (!parameter) return;
        if ("value" in r.current) {
            switch (parameter.type) {
                case ParameterTypes.InputDirectory:
                case ParameterTypes.InputFile:
                    const expandedValue = expandHomeFolder(r.current.value, cloud.homeFolder);
                    extracted[key] = {
                        source: expandedValue,
                        destination: removeTrailingSlash(expandedValue).split("/").pop()!
                    };
                    return;
                case ParameterTypes.Boolean:
                    switch (r.current.value) {
                        case "Yes":
                            extracted[key] = true;
                            return;
                        case "No":
                            extracted[key] = false;
                            return;
                        default:
                            return;
                    }
                case ParameterTypes.Integer:
                    extracted[key] = parseInt(r.current.value, 10);
                    return;
                case ParameterTypes.FloatingPoint:
                    extracted[key] = parseFloat(r.current.value);
                    return;
                case ParameterTypes.Text:
                    extracted[key] = r.current.value;
                    return;
                case ParameterTypes.SharedFileSystem:
                    extracted[key] = {fileSystemId: r.current.value};
                    return;
                case ParameterTypes.Peer:
                    extracted[key] = {jobId: r.current.value};
                    return;
            }
        } else {
            switch (parameter.type) {
                case ParameterTypes.Range:
                    const {bounds} = r.current.state;
                    extracted[key] = {min: bounds[0], max: bounds[1]};
                    return;
            }
        }
    });
    return extracted;
}

export const inCancelableState = (state: JobState) =>
    state === JobState.VALIDATED ||
    state === JobState.PREPARED ||
    state === JobState.SCHEDULED ||
    state === JobState.RUNNING;


export function validateOptionalFields(
    invocation: ApplicationInvocationDescription,
    parameters: ParameterValues
): boolean {
    const optionalErrors = [] as string[];
    const optionalParams = invocation.parameters.filter(it => it.optional && it.visible).map(it =>
        ({name: it.name, title: it.title})
    );
    optionalParams.forEach(it => {
        const {current} = parameters.get(it.name)!;
        if (current == null || !("checkValidity" in current)) return;
        if (("checkValidity" in current! && !current!.checkValidity())) optionalErrors.push(it.title);
    });

    if (optionalErrors.length > 0) {
        snackbarStore.addFailure(
            `Invalid values for ${optionalErrors.slice(0, 3).join(", ")}
                    ${optionalErrors.length > 3 ? `and ${optionalErrors.length - 3} others` : ""}`,
            5000
        );
        return false;
    }

    return true;
}

export function checkForMissingParameters(
    parameters: ExtractedParameters,
    invocation: ApplicationInvocationDescription
): boolean {
    const PT = ParameterTypes;
    const requiredParams = invocation.parameters.filter(it => !it.optional);
    const missingParameters: string[] = [];
    requiredParams.forEach(rParam => {
        const parameterValue = parameters[rParam.name];
        if (parameterValue == null) missingParameters.push(rParam.title);
        else if ([PT.Boolean, PT.FloatingPoint, PT.Integer, PT.Text].includes[rParam.type] &&
            !["number", "string", "boolean"].includes(typeof parameterValue)) {
            missingParameters.push(rParam.title);
        } else if (rParam.type === ParameterTypes.InputDirectory || rParam.type === ParameterTypes.InputFile) {
            if (!parameterValue["source"]) {
                missingParameters.push(rParam.title);
            }
        } else if (rParam.type === ParameterTypes.SharedFileSystem) {
            if (!parameterValue["fileSystemId"]) {
                missingParameters.push(rParam.title);
            }
        }
    });

    // Check missing values for required input fields.
    if (missingParameters.length > 0) {
        snackbarStore.addFailure(
            `Missing values for ${missingParameters.slice(0, 3).join(", ")}
                ${missingParameters.length > 3 ? `and ${missingParameters.length - 3} others.` : ``}`,
            5000
        );
        return false;
    }
    return true;
}
