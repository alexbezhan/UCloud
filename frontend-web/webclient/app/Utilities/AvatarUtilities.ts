export const saveAvatarQuery = "/avatar/update";
export const findAvatarQuery = "/avatar/find";
