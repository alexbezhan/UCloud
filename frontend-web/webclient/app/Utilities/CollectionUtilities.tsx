export function removeEntry<T>(arr: T[], index: number): T[] {
    return arr.slice(0, index).concat(arr.slice(index + 1));
}

export function setsDiffer<T>(s1: Set<T>, s2: Set<T>): boolean {
    if (s1.size !== s2.size) return true;
    let differ = false;
    s1.forEach(it => {
        if (!s2.has(it)) differ = true;
    });
    return differ;
}

type PrimitiveDataTypes = string | number | boolean;

/**
 * @param set
 * @param entry
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size !== set.size;
}

export function groupBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T[]> {
    const result: Record<string, T[]> = {};
    items.forEach(item => {
        const key = keySelector(item);
        const newList = result[key] ?? [];
        newList.push(item);
        result[key] = newList;
    });
    return result;
}

export function associateBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T> {
    const result: Record<string, T> = {};
    items.forEach(item => {
        const key = keySelector(item);
        result[key] = item;
    });
    return result;
}

export function takeLast<T>(items: T[], numberOfItems: number): T[] {
    return items.slice(Math.max(0, items.length - numberOfItems));
}
