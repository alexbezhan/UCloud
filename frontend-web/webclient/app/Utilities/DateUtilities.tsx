import {format} from "date-fns/esm";

// Could potentially cause issues with time if user is outside CEST
export const dateToString = (date: number) => format(date, "HH:mm dd/MM/yyyy");
