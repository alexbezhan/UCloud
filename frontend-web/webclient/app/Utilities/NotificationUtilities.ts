export const notificationsQuery = "/notifications?itemsPerPage=100";
export const readAllNotificationsQuery = "notifications/read/all";
export const readNotificationQuery = (id: number) => `notifications/read/${id}`;
