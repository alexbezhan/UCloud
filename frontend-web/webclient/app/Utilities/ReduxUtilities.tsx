import activity from "Activity/Redux/ActivityReducer";
import * as AppRedux from "Applications/Redux";
import analyses from "Applications/Redux/AnalysesReducer";
import applications from "Applications/Redux/BrowseReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import * as TaskRedux from "BackgroundTasks/redux";
import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import fileInfo from "Files/Redux/FileInfoReducer";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import uploader from "Uploader/Redux/UploaderReducer";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    activity,
    dashboard,
    analyses,
    applications,
    header,
    status,
    sidebar,
    uploader,
    notifications,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    fileInfo,
    hookStore,
    ...AppRedux.reducers,
    avatar,
    loading,
    tasks: TaskRedux.reducer,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
