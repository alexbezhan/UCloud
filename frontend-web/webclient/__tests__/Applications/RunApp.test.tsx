import {createMemoryHistory} from "history";
import * as React from "react";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router";
import {create} from "react-test-renderer";
import {ThemeProvider} from "styled-components";
import RunApp from "../../app/Applications/Run";
import theme from "../../app/ui-components/theme";
import {store} from "../../app/Utilities/ReduxUtilities";

describe("RunApp component", () => {
    test("Mount", () => {
        expect(create(
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <MemoryRouter>
                        <RunApp
                            match={{
                                params: {appName: "appName", appVersion: "appVersion"},
                                isExact: true,
                                path: "",
                                url: ""
                            }}
                            location={{search: ""}}
                            history={createMemoryHistory()}
                        />
                    </MemoryRouter>
                </ThemeProvider>
            </Provider>
        ).toJSON()
        ).toMatchSnapshot();
    });

});
