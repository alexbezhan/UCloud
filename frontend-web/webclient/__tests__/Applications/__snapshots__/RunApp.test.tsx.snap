// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RunApp component Mount 1`] = `
.c0 {
  cursor: auto;
  margin-left: 190px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  background-color: #fff;
}

.c1 {
  cursor: auto;
  padding-top: 14px;
  padding-right: 14px;
}

.c2 {
  cursor: auto;
}

.c3 {
  width: 18px;
  height: 18px;
  margin: 20px auto;
  -webkit-animation-name: cIivWO;
  animation-name: cIivWO;
}

.c3 > svg path:nth-child(1) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.04s;
  animation-delay: -0.04s;
}

.c3 > svg path:nth-child(2) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.08s;
  animation-delay: -0.08s;
}

.c3 > svg path:nth-child(3) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.12s;
  animation-delay: -0.12s;
}

.c3 > svg path:nth-child(4) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.16s;
  animation-delay: -0.16s;
}

.c3 > svg path:nth-child(5) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.2s;
  animation-delay: -0.2s;
}

.c3 > svg path:nth-child(6) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.24s;
  animation-delay: -0.24s;
}

.c3 > svg path:nth-child(7) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.28s;
  animation-delay: -0.28s;
}

.c3 > svg path:nth-child(8) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.32s;
  animation-delay: -0.32s;
}

.c3 > svg path:nth-child(9) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.36s;
  animation-delay: -0.36s;
}

.c3 > svg path:nth-child(10) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.4s;
  animation-delay: -0.4s;
}

.c3 > svg path:nth-child(11) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.44s;
  animation-delay: -0.44s;
}

.c3 > svg path:nth-child(12) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.48s;
  animation-delay: -0.48s;
}

.c3 > svg path:nth-child(13) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.52s;
  animation-delay: -0.52s;
}

.c3 > svg path:nth-child(14) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.56s;
  animation-delay: -0.56s;
}

.c3 > svg path:nth-child(15) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.6s;
  animation-delay: -0.6s;
}

.c3 > svg path:nth-child(16) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.64s;
  animation-delay: -0.64s;
}

.c3 > svg path:nth-child(17) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.68s;
  animation-delay: -0.68s;
}

.c3 > svg path:nth-child(18) {
  -webkit-animation: cIivWO 0.72s linear infinite;
  animation: cIivWO 0.72s linear infinite;
  -webkit-animation-delay: -0.72s;
  animation-delay: -0.72s;
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c2 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c2 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c2 {
    display: none;
  }
}

<div
  className="c0"
  cursor="auto"
>
  <div
    className="c1"
    cursor="auto"
  >
    <div
      className="c2"
      cursor="auto"
    />
    <div
      className="c3"
      height={18}
      width={18}
    >
      <svg
        height={18}
        id="svg"
        version="1.1"
        viewBox="0 0 220 220"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <path
          d="M80.296,0.12l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
        />
        <path
          d="M149.73,69.554l-14.673,-54.761l-40.088,40.088l54.761,14.673Z"
        />
        <path
          d="M189.818,29.466l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
        />
        <path
          d="M204.491,84.228l-14.673,-54.762l-40.088,40.088l54.761,14.674Z"
        />
        <path
          d="M164.403,124.316l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
        />
        <path
          d="M219.165,138.989l-14.674,-54.761l-40.088,40.088l54.762,14.673Z"
        />
        <path
          d="M179.077,179.077l-14.674,-54.761l54.762,14.673l-40.088,40.088Z"
        />
        <path
          d="M124.316,164.403l54.761,14.674l-14.674,-54.761l-40.087,40.087Z"
        />
        <path
          d="M138.989,219.165l40.088,-40.088l-54.761,-14.674l14.673,54.762Z"
        />
        <path
          d="M84.228,204.491l54.761,14.674l-14.673,-54.762l-40.088,40.088Z"
        />
        <path
          d="M69.554,149.73l14.674,54.761l40.088,-40.088l-54.762,-14.673Z"
        />
        <path
          d="M29.466,189.818l54.762,14.673l-14.674,-54.761l-40.088,40.088Z"
        />
        <path
          d="M14.793,135.057l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
        />
        <path
          d="M54.881,94.969l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
        />
        <path
          d="M0.12,80.296l14.673,54.761l40.088,-40.088l-54.761,-14.673Z"
        />
        <path
          d="M40.208,40.208l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
        />
        <path
          d="M94.969,54.881l-54.761,-14.673l14.673,54.761l40.088,-40.088Z"
        />
        <path
          d="M80.296,0.12l-40.088,40.088l54.761,14.673l-14.673,-54.761Z"
        />
      </svg>
    </div>
  </div>
</div>
`;
